[OF-SDK#825] Add custom children component

vaszig opened the pull request:
Closes open-formulieren/open-forms-sdk#825 partly

### Requires:

- [x] new version of formio-builder
- [x] open-formulieren/open-forms-sdk#847 to be merged

**Changes**

- Added the custom children component

**Checklist**

Check off the items that are completed or not relevant.

- Impact on features

  - [x] Checked copying a form
  - [x] Checked import/export of a form
  - [x] Config checks in the configuration overview admin page
  - [x] Problem detection in the admin email digest is handled

- Release management

  - [x] I have labelled the PR as "needs-backport" accordingly

- I have updated the translations assets (you do NOT need to provide translations)

  - [x] Ran `./bin/makemessages_js.sh`
  - [x] Ran `./bin/compilemessages_js.sh`

- Dockerfile/scripts

  - [x] Updated the Dockerfile with the necessary scripts from the `./bin` folder

- Commit hygiene

  - [x] Commit messages refer to the relevant Github issue
  - [x] Commit messages explain the "why" of change, not the how

vaszig commented on src/openforms/formio/components/custom.py:
```diff
@@ -769,6 +770,135 @@ def as_json_schema(component: Component) -> JSONObject:
+    firstNames = serializers.CharField(
+        label=_("firstNames"),
+        help_text=_("The first names of the child"),
+        required=False,
+        allow_blank=True,
```
Yes, it's required when the user adds a child manually (like the partners, only the BSN is required).

vaszig commented on src/openforms/formio/components/custom.py:
```diff
@@ -769,6 +770,135 @@ def as_json_schema(component: Component) -> JSONObject:
+    dateOfBirth = serializers.DateField(
+        label=_("date of birth"),
+        help_text=_("The date of birth of the child"),
+        required=False,
```
Same as above.

vaszig commented on src/openforms/submissions/tests/test_submission_step_validate.py:
```diff
@@ -936,3 +936,143 @@ def test_partners_component_fails_when_data_is_tampered(self):
+        response = self.client.post(
+            endpoint,
+            {
+                "data": {
+                    "children": [
+                        {
+                            "bsn": "999970409",
+                            "affixes": "van",
+                            "initials": "P.",
+                            "lastName": "Paassen",
+                            "firstNames": "Pero",
+                            "dateOfBirth": "2023-02-01",
+                            "dateOfBirthPrecision": "date",
+                            "selected": False,
+                            "__addedManually": False,
+                        },
+                    ]
+                }
+            },
```
This is the data coming from the frontend. I will add a comment.

robinmolen replied:
But then it should only contain `bsn`, `firstNames`, `dateOfBirth`, `selected` and `__addedManually` (and perhaps `__id`?) right?

```
"data": {
  "children": [
    {
     "bsn": "999970409",
      "firstNames": "Pero",
      "dateOfBirth": "2023-02-01",
      "selected": False,
      "__addedManually": False,
    },
  ]
}
```

vaszig commented:
@robinmolen FYI I have added the comments we discussed and besides that I added one more missing test for the valid data (without all the keys that the frontend returns) with only the required keys

https://github.com/open-formulieren/open-forms/pull/5438/files#diff-38bbaf27753ad6f2523d8168cf3546248409e1b4be515ef1f9effdfafef124b1R10

vaszig commented:
@robinmolen While I was trying to implement the second step of the children component flow, I found a couple of problems and things that I would like to improve so that's why I pushed one more commit. You can see in the description of the commit the bullet points.

sergei-maertens commented on src/openforms/formio/components/custom.py:
```diff
@@ -770,6 +771,156 @@ def as_json_schema(component: Component) -> JSONObject:
+            source=FormVariableSources.user_defined,
+            prefill_plugin=FM_PLUGIN_IDENTIFIER,
+            prefill_options__mutable_data_form_variable=component_key,
+            form=submission.form.id,
```
```suggestion
            form=submission.form,
```
it's more pythonic to reason with the model instances than their underlying database ID

vaszig replied:
Apologies, it was there before our last discussion and I forgot about it.

sergei-maertens commented on src/openforms/formio/components/custom.py:
```diff
@@ -770,6 +771,156 @@ def as_json_schema(component: Component) -> JSONObject:
+            # We also receive the boolean `selected` which is set on the frontend and
+            # therefore should not be part of the data for validation (in combination
+            # with the retrieved data from the external source)
+            for child in children:
+                child.pop("selected", None)
```
I think it's a bit smelly that a `validate` function performs side effects by mutating its argument. To keep a pure function, it'd probably be best to `copy.deepcopy(children)`, otherwise weird things will happen if we will use the validation serializer output as data to save in variables (something we probably want to do in the long term)

sergei-maertens commented on src/openforms/formio/components/custom.py:
```diff
@@ -770,6 +771,156 @@ def as_json_schema(component: Component) -> JSONObject:
+                for child in prefill_data[fm_immutable_variable.key]
+            ]
+
+            if not self.validate_selected_children(children):
+                raise serializers.ValidationError(_("Invalid data."))
```
I would probably call `self.validate_selected_children(...)` and put the `raise` in that method, for consistency with other validation functions. Or, raise here, but rename the method to `def check_selected_key_tampering(...)` or something like that that makes it more clear that the return value is relevant

sergei-maertens commented on src/openforms/formio/components/custom.py:
```diff
@@ -770,6 +771,156 @@ def as_json_schema(component: Component) -> JSONObject:
+                    "firstNames": {"type": "string"},
+                    "dateOfBirth": {"type": "string", "format": "date"},
+                },
+                "additionalProperties": False,
```
is that true? will we not send extra properties like `affixes` and `lastName` etc.?
